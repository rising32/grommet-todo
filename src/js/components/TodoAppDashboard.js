import React, { Component, PropTypes } from 'react';

import Box from 'grommet/components/Box';
import Button from 'grommet/components/Button';
import Heading from 'grommet/components/Heading';
import Meter from 'grommet/components/Meter';
import Section from 'grommet/components/Section';
import List from 'grommet/components/List';
import ListItem from 'grommet/components/ListItem';
import Value from 'grommet/components/Value';

import Status from 'grommet/components/icons/Status';
import CloseIcon from 'grommet/components/icons/base/Close';

import TodoAddTaskForm from './TodoAddTaskForm';

function getLabel(label, value, colorIndex) {
  return { label, value, colorIndex };
}

class DeleteTaskButton extends Component {
  constructor() {
    super();
    this._onClick = this._onClick.bind(this);
  }

  _onClick() {
    const { task, onDelete } = this.props;
    onDelete(task);
  }

  render() {
    const { task } = this.props;
    return (
      <Button plain={true}
        onClick={this._onClick}
        icon={<CloseIcon />}
        a11yTitle={`Delete ${task.item} task`} />
    );
  }
}

DeleteTaskButton.propTypes = {
  task: PropTypes.object.isRequired,
  onDelete: PropTypes.func.isRequired
};

export default class TodoAppDashboard extends Component {

  constructor() {
    super();

    this._onRequestForAdd = this._onRequestForAdd.bind(this);
    this._onRequestForAddClose = this._onRequestForAddClose.bind(this);
    this._onRequestForDelete = this._onRequestForDelete.bind(this);
    this._onAddTask = this._onAddTask.bind(this);

    this.state = {
      tasks: [],
      addTask: false
    };
  }

  _onRequestForAdd() {
    this.setState({ addTask: true });
  }

  _onRequestForAddClose() {
    this.setState({ addTask: false });
  }

  _onRequestForDelete(task) {
    const { tasks } = this.state;
    const index = this.state.tasks.indexOf(task);
    tasks.splice(index, 1);
    this.setState({ tasks });
  }

  _onAddTask(task) {
    const tasks = this.state.tasks;
    tasks.push(task);
    this.setState({ tasks, addTask: false });
  }

  _onMeterActive(index) {
    this.setState({ index });
  }

  render() {
    const tasksMap = {
      critical: 0,
      ok: 0,
      warning: 0
    };

    const tasks = this.state.tasks.map((task, index) => {
      tasksMap[task.status] += 1;

      let separator;
      if (index === 0) {
        separator = 'horizontal';
      }
      return (
        <ListItem key={`task_${index}`} justify='between'
          separator={separator} responsive={false}>
          <Box direction='row' responsive={false}
            pad={{ between: 'small' }}>
            <Status value={task.status} size='small' />
            <span>{task.label}</span>
          </Box>
          <DeleteTaskButton task={task} onDelete={this._onRequestForDelete} />
        </ListItem>
      );
    }, this);

    let addTask;
    if (this.state.addTask) {
      addTask = (
        <TodoAddTaskForm onClose={this._onRequestForAddClose}
          onSubmit={this._onAddTask} />
      );
    }

    const series = [
      getLabel('Past Due', tasksMap.critical, 'critical'),
      getLabel('Due Soon', tasksMap.warning, 'warning'),
      getLabel('Done', tasksMap.ok, 'ok')
    ];

    let value;
    let label;
    if (this.state.index >= 0) {
      value = series[this.state.index].value;
      label = series[this.state.index].label;
    } else {
      value = 0;
      series.forEach(serie => value += serie.value);
      label = 'Total';
    }

    return (
      <Section primary={true} flex={true}>
        <Box direction='row'>
          <Box basis='1/3' align='center'>
            <Meter series={series} type='circle' label={false}
              onActive={this._onMeterActive} />
            <Box direction='row' justify='between' align='center'
              responsive={false}>
              <Value value={value} units='Tasks' align='center' label={label} />
            </Box>
          </Box>
          <Box basis='2/3' pad='medium'>
            <Heading tag='h3'>My Tasks</Heading>
            <List>
              {tasks}
            </List>
            <Box pad={{ vertical: 'large' }} align='start'>
              <Button label='Add Task' primary={true}
                onClick={this._onRequestForAdd} />
            </Box>
          </Box>
        </Box>
        {addTask}
      </Section>
    );
  }
}
