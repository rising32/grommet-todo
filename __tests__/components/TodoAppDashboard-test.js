import React from 'react';
import TodoAppDashboard from '../../src/js/components/TodoAppDashboard';
import renderer from 'react-test-renderer';

test('TodoAppDashboard renders', () => {
  const component = renderer.create(
    <TodoAppDashboard />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
