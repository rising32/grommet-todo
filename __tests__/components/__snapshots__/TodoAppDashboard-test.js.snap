exports[`test TodoAppDashboard renders 1`] = `
<section
  className="grommetux-box grommetux-box--direction-column grommetux-box--responsive grommetux-box--flex grommetux-box--pad-vertical-medium grommetux-section"
  id={undefined}
  onClick={undefined}
  role={undefined}
  style={Object {}}
  tabIndex={undefined}>
  <a
    aria-hidden="true"
    className="grommetux-skip-link-anchor"
    id="skip-link-main_content"
    tabIndex="-1">
    Main Content
  </a>
  <div
    className="grommetux-box grommetux-box--direction-row grommetux-box--responsive grommetux-box--pad-none"
    id={undefined}
    onClick={undefined}
    role={undefined}
    style={Object {}}
    tabIndex={undefined}>
    <div
      className="grommetux-box grommetux-box--direction-column grommetux-box--align-center grommetux-box--responsive grommetux-box--basis-1-3 grommetux-box--pad-none"
      id={undefined}
      onClick={undefined}
      role={undefined}
      style={Object {}}
      tabIndex={undefined}>
      <div
        className="grommetux-meter grommetux-meter--circle grommetux-meter--count-3">
        <div
          className="grommetux-meter__value-container">
          <div
            className="grommetux-meter__graphic-container">
            <svg
              aria-label="Circle Meter. , Value: 0, Max: 100"
              className="grommetux-meter__graphic"
              height={192}
              onBlur={[Function]}
              onFocus={[Function]}
              preserveAspectRatio="xMidYMid meet"
              role="group"
              tabIndex="0"
              viewBox="0 0 192 192"
              width={192}>
              <g
                className="grommetux-meter__tracks">
                <g>
                  <path
                    className="grommetux-meter__slice"
                    d="M 96.0146607656 179.9999987206 A 84.0000000000 84.0000000000 0 1 0 96.0000000000 180.0000000000"
                    data-index={0}
                    onBlur={[Function]}
                    onFocus={[Function]} />
                  <path
                    className="grommetux-meter__hot"
                    d="M 96.0146607656 179.9999987206 A 84.0000000000 84.0000000000 0 1 0 96.0000000000 180.0000000000"
                    fill="none"
                    onClick={undefined}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]} />
                </g>
                <g>
                  <path
                    className="grommetux-meter__slice"
                    d="M 96.0104719755 155.9999990861 A 60.0000000000 60.0000000000 0 1 0 96.0000000000 156.0000000000"
                    data-index={1}
                    onBlur={[Function]}
                    onFocus={[Function]} />
                  <path
                    className="grommetux-meter__hot"
                    d="M 96.0104719755 155.9999990861 A 60.0000000000 60.0000000000 0 1 0 96.0000000000 156.0000000000"
                    fill="none"
                    onClick={undefined}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]} />
                </g>
                <g>
                  <path
                    className="grommetux-meter__slice"
                    d="M 96.0062831853 131.9999994517 A 36.0000000000 36.0000000000 0 1 0 96.0000000000 132.0000000000"
                    data-index={2}
                    onBlur={[Function]}
                    onFocus={[Function]} />
                  <path
                    className="grommetux-meter__hot"
                    d="M 96.0062831853 131.9999994517 A 36.0000000000 36.0000000000 0 1 0 96.0000000000 132.0000000000"
                    fill="none"
                    onClick={undefined}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]} />
                </g>
              </g>
              <g
                className="grommetux-meter__values">
                <g
                  aria-label="0"
                  role="img">
                  <path
                    className="grommetux-meter__slice grommetux-color-index-critical"
                    d="M 96.0000000000 180.0000000000 A 84.0000000000 84.0000000000 0 0 0 96.0000000000 180.0000000000"
                    data-index={0}
                    onBlur={[Function]}
                    onFocus={[Function]} />
                  <path
                    className="grommetux-meter__hot"
                    d="M 96.0000000000 180.0000000000 A 84.0000000000 84.0000000000 0 0 0 96.0000000000 180.0000000000"
                    fill="none"
                    onClick={undefined}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]} />
                </g>
                <g
                  aria-label="0"
                  role="img">
                  <path
                    className="grommetux-meter__slice grommetux-color-index-warning"
                    d="M 96.0000000000 156.0000000000 A 60.0000000000 60.0000000000 0 0 0 96.0000000000 156.0000000000"
                    data-index={1}
                    onBlur={[Function]}
                    onFocus={[Function]} />
                  <path
                    className="grommetux-meter__hot"
                    d="M 96.0000000000 156.0000000000 A 60.0000000000 60.0000000000 0 0 0 96.0000000000 156.0000000000"
                    fill="none"
                    onClick={undefined}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]} />
                </g>
                <g
                  aria-label="0"
                  role="img">
                  <path
                    className="grommetux-meter__slice grommetux-color-index-ok"
                    d="M 96.0000000000 132.0000000000 A 36.0000000000 36.0000000000 0 0 0 96.0000000000 132.0000000000"
                    data-index={2}
                    onBlur={[Function]}
                    onFocus={[Function]} />
                  <path
                    className="grommetux-meter__hot"
                    d="M 96.0000000000 132.0000000000 A 36.0000000000 36.0000000000 0 0 0 96.0000000000 132.0000000000"
                    fill="none"
                    onClick={undefined}
                    onMouseOut={[Function]}
                    onMouseOver={[Function]} />
                </g>
              </g>
            </svg>
          </div>
        </div>
      </div>
      <div
        className="grommetux-box grommetux-box--direction-row grommetux-box--justify-between grommetux-box--align-center grommetux-box--pad-none"
        id={undefined}
        onClick={undefined}
        role={undefined}
        style={Object {}}
        tabIndex={undefined}>
        <div
          className="grommetux-value grommetux-value--align-center">
          <div
            className="grommetux-value__annotated">
            <span
              className="grommetux-value__value">
              0
            </span>
            <span
              className="grommetux-value__units">
              Tasks
            </span>
          </div>
          <span
            className="grommetux-value__label">
            Total
          </span>
        </div>
      </div>
    </div>
    <div
      className="grommetux-box grommetux-box--direction-column grommetux-box--responsive grommetux-box--basis-2-3 grommetux-box--pad-medium"
      id={undefined}
      onClick={undefined}
      role={undefined}
      style={Object {}}
      tabIndex={undefined}>
      <h3
        className="grommetux-heading">
        My Tasks
      </h3>
      <ul
        className="grommetux-list"
        role="list" />
      <div
        className="grommetux-box grommetux-box--direction-column grommetux-box--align-start grommetux-box--responsive grommetux-box--pad-vertical-large"
        id={undefined}
        onClick={undefined}
        role={undefined}
        style={Object {}}
        tabIndex={undefined}>
        <button
          aria-label={undefined}
          className="grommetux-button grommetux-button--primary"
          disabled={false}
          href={undefined}
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onMouseDown={[Function]}
          onMouseUp={[Function]}
          type="button">
          <span
            className="grommetux-button__label">
            Add Task
          </span>
        </button>
      </div>
    </div>
  </div>
</section>
`;
